File: examples/bar.js
import BarBar from 'gi://BarBar';
import Gtk from 'gi://Gtk?version=4.0';
import Systray from 'gi://Statusnotifier';
import GObject from 'gi://GObject';
import GLib from 'gi://GLib';

let manager = [];

function newscreen(head, app) {

  // head.connect('notify::resolution-height', (source, paramSpec) => {
  //   print(`The property ${paramSpec.name} has changed to: ${source.resolutionHeight}`);
  // });
  print(head.get_resolution_width());
  if (head.get_name() == "eDP-1") {
    let builder = BarBar.default_builder("barbar/config.ui");

    let list = builder.get_objects();

    for (let obj of list) {
      if (obj instanceof Gtk.Window) {
        app.add_window(obj);
        obj.present();
      } else if (obj instanceof BarBar.Sensor) {
        manager.push(obj);
        obj.start();
      } else if (obj instanceof Gtk.ScaleButton) {
        let minus = obj.get_minus_button();
        let plus = obj.get_plus_button();
        minus.set_visible(false);
        plus.set_visible(false);
      }
    }
  }
}

Gtk.init();
BarBar.init();
// Systray doesn't have a init function
GObject.type_ensure(Systray.Systray);

function activate(app) {
  BarBar.default_style_provider("barbar/style.css");

  let context = GLib.MainContext.default();
  let outputs = new BarBar.OutputManager();

  outputs.connect("new-head", (_, head) => {
    newscreen(head, app);
  });
  outputs.run();

}

let app = new Gtk.Application({ application_id: "com.github.barbar" });
app.connect("activate", activate);
app.run(ARGV);

File: examples/bar.lua
-- GTK4 only works with the git version of lgi. Sadly lgi hasn't made a release in years.
-- So you need the git version.

-- For GTK4 Layer Shell to get linked before libwayland-client we must explicitly load it before importing with lgi
-- If you use luajit you can do this.
--
-- local ffi = require("ffi")
-- ffi.cdef[[
--     void *dlopen(const char *filename, int flags);
-- ]]
-- ffi.C.dlopen("libgtk4-layer-shell.so", 0x101)

-- Now open the library normally with LGI
local lgi = require("lgi")
local Gtk = lgi.require("Gtk", "4.0")
local BarBar = lgi.require("BarBar", "1.0")
local GLib = lgi.require("GLib")

Gtk.init()
BarBar.init()

local manager = {}
local ui = GLib.get_user_config_dir() .. "/barbar/config.ui"

local app = Gtk.Application({
	application_id = "com.github.barbar",
})
app.on_activate = function()
	local builder = Gtk.Builder()

	builder:add_from_file(ui)

	local list = builder:get_objects()

	for _, obj in pairs(list) do
		if Gtk.Window:is_type_of(obj) then
			app:add_window(obj)
			obj:present()
		elseif BarBar.Sensor:is_type_of(obj) then
			table.insert(manager, obj)
			obj:start()
		end
	end
end
app:run()
